LIT-3920 - wrapped-keys - Batch generate private keys & optionally sign messages in a single operation

MaximusHaximus opened the pull request:
# Description

This PR adds a new API function to `wrapped-keys` and corresponding LITAction to `wrapped-keys-lit-actions` facilitates the execution of multiple operations -- previously requiring executing *4* independent actions -- in a single API call to a single LIT action execution.

#### Noteworthy changes in this PR:
- Refactored existing LIT actions to allow composition of existing generation and message signing functionality into the new LIT action, while using the same code in the existing LIT actions that individually generate keys or sign messages with keys for a specific network (evm or solana)
- Added a new `LIT_ACTION_CID_REPOSITORY_COMMON` constant, and a new `litActionCodeRepositoryCommon` constant to `wrapped-keys`; 'common' LIT actions are not network-specific, so these repositories do not have multiple entries keyed by `Network` (evm or solana) under each LIT action entry; instead there is a single entry per LIT action which provides functionality for multiple networks.

## `wrapped-keys`
#### New  functions
##### `api.batchGeneratePrivateKeys()`
- Takes an array of `actions`, where each action can define either network (evm or solana), and an optional message to sign and return the signature of
##### `config.setLitActionsCodeCommon()`
- Allows injection of a LIT actions code repository containing 'common' LIT actions that are not defined per-network
#### New types
##### `BatchGeneratePrivateKeysResult` / `BatchGeneratePrivateKeysActionResult`
##### `BatchGeneratePrivateKeysParams` / `GeneratePrivateKeyAction`
##### `LitActionCodeRepositoryCommon`

## `wrapped-keys-lit-actions`
#### New LIT action
##### `batchGenerateEncryptedKeys()`
- Facilitates generating `n` new keys _and_ optionally signing a message for each key, and returning the results of all generated keys and signed messages in a _single execution of one LIT action_


## Type of change

- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

# How Has This Been Tested?


- [X] Ran all wrapped-keys localtests and added a new one to test the new batch generate/sign LIT action
- [X] Wrote and executed custom tests to verify that existing LIT actions continue to function identically to how they did before the internal refactors done in this PR

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

MaximusHaximus commented on local-tests/setup/tinny-environment.ts:
```diff
@@ -460,6 +460,9 @@ export class TinnyEnvironment {
       throw new Error(`Failed to send funds to ${walletAddress}: ${e}`);
+    } finally {
+      // @ts-expect-error We don't have a user, but this works
+      this.releasePrivateKeyFromUser({ privateKey });
```
Fixes the infinite hang during test run when you only have a couple of private keys in scope in your environment var

MaximusHaximus commented on local-tests/test.ts:
```diff
@@ -104,6 +104,12 @@ import { testFailImportWrappedKeysWithExpiredSessionSig } from './tests/wrapped-
+import { setLitActionsCodeToLocal } from './tests/wrapped-keys/util';
+
+// Use the current LIT action code to test against
+setLitActionsCodeToLocal();
```
This makes our tinny tests use the locally built code for all LIT actions in wrapped-keys instead of running them against the CIDs in `wrapped-keys` LIT_ACTION_CID_REPOSITORY

DashKash54 replied:
Since the idea is that users use the IPFS id shouldn't the tests reflect that?

MaximusHaximus replied:
I guess it depends on the users; some people will be using the code directly, others the CIDs... what I want to do is run the entire suite against the cids and then again against the current code ideally, but see other comment re: doing so without jest.

The reason this is here right now is because we have a chicken and egg problem - in-progress code in PRs isn't in IPFS, and pinning repeatedly to IPFS as we iterate on a lit action would result in a ton of churn and manual updates right now.

We could make this code always use IPFS CIDs, but it would require that all actions we want to test would exist in IPFS before we can test against them -- and we want to run tests while developing the actions where they will be changing often.  

In this PR, I wanted to be sure tests pass with the _current code_ rather than previous CIDs -- once we have confirmed all of the LIT actions code here is GTG and published all of the lit actions to IPFS, I can update the CIDs in `wrapped-keys` and remove this code so the tests run against CIDs again

MaximusHaximus replied:
:+1: I agree @FedericoAmura <3 Ideally, all tests would run against the data in the same commit by _default_, and if we want to test against prior versions we'd explicitly do _that_.  The main thing is just getting the CID repo in the same commit to always match the current source code in that commit... I think we can do that with a husky hook script that runs on-commit

DashKash54 replied:
> husky hook script that runs on-commit

Good idea

MaximusHaximus commented on local-tests/tests/wrapped-keys/testFailEthereumSignTransactionWrappedKeyWithInvalidParam.ts:
```diff
@@ -70,11 +70,7 @@ export const testFailEthereumSignTransactionWrappedKeyWithInvalidParam = async (
-          'Error executing the Signing Lit Action: Error: When signing transaction- invalid hexlify value'
-        )
-      ) {
+      if (e.message.includes('invalid hexlify value')) {
```
Made this less specific since the exact string isn't important and this is now being properly prefixed consistently due to refactors in error handling in the LA in this PR

MaximusHaximus commented on local-tests/tests/wrapped-keys/testFailEthereumSignTransactionWrappedKeyWithMissingParam.ts:
```diff
@@ -66,11 +66,7 @@ export const testFailEthereumSignTransactionWrappedKeyWithMissingParam = async (
-          'Error executing the Signing Lit Action: Error: Missing required field: toAddress'
-        )
-      ) {
+      if (e.message.includes('Missing required field: toAddress')) {
```
Made this less specific since the exact string isn't important and this is now being properly prefixed consistently due to refactors in error handling in the LA in this PR

MaximusHaximus commented on local-tests/tests/wrapped-keys/testSignTransactionWithSolanaEncryptedKey.ts:
```diff
@@ -125,7 +126,7 @@ export const testSignTransactionWithSolanaEncryptedKey = async (
     // console.log(confirmation); // { context: { slot: 321490379 }, value: { err: null } }
 
-    const signatureBuffer = Buffer.from(signedTx, 'base64');
+    const signatureBuffer = Buffer.from(ethers.utils.base58.decode(signedTx));
```
This test was failing if using the LIT action code merged since #652, but passed because we were using the CID to run the tests.  Now that we run the tests against the current, in-repo LIT actions, it needed to be updated to match the changes made in #652.

DashKash54 replied:
Shouldn't we use IPFS Cid in the tests too? https://github.com/LIT-Protocol/js-sdk/pull/675#discussion_r1792648904

MaximusHaximus replied:
Once all the actions have been reviewed and we've published/pinned them to IPFS, we can remove the injection of the LIT action code here.  Right now it's using injected code so that we can be sure this commit is sane and functional even though it's not in IPFS yet (see other comment)

MaximusHaximus commented on local-tests/tests/wrapped-keys/util.ts:
```diff
@@ -1,7 +1,50 @@
+  batchGenerateEncryptedKeys: '',
+};
+
+const emptyLitActionRepository: LitActionCodeRepository = {
```
Sadly we can't run all of the tests twice (once w/ cids and once w/ actions) because this is a global mutation and it would require runInBand=true to avoid inconsistent behaviour.  When we migrate to Jest, I'll make that happen (since Jest sandboxes suites)

DashKash54 replied:
Run them sequentially?

MaximusHaximus replied:
We'd have to configure the entire suite to always run _all_ tests to `runInBand` the way `local-tests` is currently designed; since we're already in the midst of migrating to Jest and it will be able to run them in parallel it didn't seem like the effort was justified to enhance the existing local-tests code to allow subsets of tests to run in band

MaximusHaximus commented on packages/wrapped-keys-lit-actions/src/index.ts:
```diff
@@ -1,12 +1,16 @@
 import * as generateEncryptedEthereumPrivateKey from './generated/ethereum/generateEncryptedEthereumPrivateKey';
-import * as signMessageWithEthereumEncryptedKey from './generated/ethereum/signMessageWithEthereumEncryptedKey';
-import * as signTransactionWithEthereumEncryptedKey from './generated/ethereum/signTransactionWithEthereumEncryptedKey';
+import * as signMessageWithEthereumEncryptedKey from './generated/ethereum/signMessageWithEncryptedEthereumKey';
```
Renamed the files for least-to-most-specific naming convention. Exports remain named the same until we ship a breaking change, then we'll make those consistent with the filenames.

MaximusHaximus commented on packages/wrapped-keys-lit-actions/src/lib/common/exportPrivateKey.js:
```diff
@@ -13,31 +16,21 @@ const { removeSaltFromDecryptedKey } = require('../utils');
     Lit.Actions.setResponse({ response: privateKey });
   } catch (err) {
-    Lit.Actions.setResponse({ response: err.message });
+    Lit.Actions.setResponse({ response: `Error: ${err.message}` });
```
New structure ensures all errors thrown from the LA result in `Error:` prefix in the response.

DashKash54 replied:
Since these are Lit Action errors should we prefix the error msg with "Lit Action Error" or the `err.message` displays that already?

MaximusHaximus replied:
In order to get the error, it has to be parsed from a node response -- so the SDK throws a specific type of error indicating that it was an error running a request against a LIT node and the stack trace from the SDK indicates it comes from handling a node response

MaximusHaximus commented on packages/wrapped-keys-lit-actions/src/lib/ethereum/internal/signTransaction.js:
```diff
@@ -0,0 +1,141 @@
+      throw new Error('Missing required field: chainId');
+    }
+
+    return {
```
Moved hexlify call to this so that we get errors about invalid parsing before running the rest of the LA code like fetching nonces etc.

MaximusHaximus commented on packages/wrapped-keys/src/lib/api/batch-generate-private-keys.ts:
```diff
@@ -0,0 +1,73 @@
+    pkpSessionSigs,
+  });
+
+  return Promise.all(
```
If any of these calls fails, the entire execution will exit even if the LA was successful and we only need to re-try storing the results using `storePrivateKey()`. A retry on network errors would be good to add at some point since the alternative is that the consumer re-runs the entire process from the top again, even though we might have valid signatures and ciphertexts in scope.

DashKash54 replied:
Cool we plan to update this later with a retry and NOT now?

MaximusHaximus replied:
:+1: yes, just making a note about the failure mode

MaximusHaximus commented on packages/wrapped-keys/src/lib/api/utils.ts:
```diff
@@ -0,0 +1,12 @@
+import { NETWORK_EVM, NETWORK_SOLANA } from '../constants';
+import { KeyType, Network } from '../types';
+
+export function getKeyTypeFromNetwork(network: Network): KeyType {
```
Extracted this to be re-used in both generate codepaths

MaximusHaximus commented on packages/wrapped-keys/src/lib/lit-actions-client/types.ts:
```diff
@@ -36,3 +42,8 @@ export type LitActionCodeRepository = Readonly<
   Record<LitActionType, LitActionCodeRepositoryEntryInput>
 >;
+
+export type LitActionCodeRepositoryCommon = Record<LitActionTypeCommon, string>;
```
New common repository just has string per key, no { evm, solana } objects since common action(s) work on multiple networks.

MaximusHaximus commented on packages/wrapped-keys/src/lib/types.ts:
```diff
@@ -134,6 +134,32 @@ export interface ExportPrivateKeyResult {
+ * @extends BaseApiParams
+ */
+export type BatchGeneratePrivateKeysParams = BaseApiParams & {
+  actions: GeneratePrivateKeyAction[];
```
Can generate `n` of any combination of solana or evm keys.

DashKash54 commented on packages/wrapped-keys-lit-actions/src/lib/common/batchGenerateEncryptedKeys.js:
A user can chose to only generate multiple keys without being necessitated to sign a message with them?

MaximusHaximus replied:
Yes, they can omit the `messageToSign` (it's optional) and just get an array of generated keys back with no message signatures :+1: :)

DashKash54 commented on packages/wrapped-keys-lit-actions/src/lib/common/batchGenerateEncryptedKeys.js:
```diff
@@ -0,0 +1,159 @@
+  const { network, generateKeyParams } = action;
+  const messageToSign = action.signMessageParams?.messageToSign;
+
+  const ethereumKey = generateEthereumPrivateKey();
```
Private key should be generated inside only a single node as this will create a different privateKey on each node! @MaximusHaximus

MaximusHaximus replied:
:+1: https://github.com/LIT-Protocol/js-sdk/pull/675#discussion_r1792721090

DashKash54 commented on packages/wrapped-keys-lit-actions/src/lib/common/batchGenerateEncryptedKeys.js:
```diff
@@ -0,0 +1,159 @@
+        JSON.stringify(
+          await encryptPrivateKey({
+            accessControlConditions,
+            publicKey: ethereumKey.publicKey,
```
Why is the pubkey required in the encrypt function since we only encrypt the private key? If we wanna return it in the same object we can just add it separately in the stringify?

MaximusHaximus replied:
Because getting the data out of the runOnce requires that we `stringify()` the result (or it comes back as [Object Object]), injecting publicKey back into the result of this function required that I `JSON.parse()` the result, then spread the parsed resulting object back into a new object and inject publicKey into it, only to then JSON.stringify() it back again on the very next line -- very inefficient and tricky to reason about.

This way I get to avoid adding another JSON.parse() to the calling code in the existing network-specific actions :+1: :)

DashKash54 replied:
No longer valid, we can change this later to move `pubKey` outside the `encryptPrivateKey`

DashKash54 commented on packages/wrapped-keys-lit-actions/src/lib/common/batchGenerateEncryptedKeys.js:
```diff
@@ -0,0 +1,159 @@
+        )
+    ),
+    messageToSign
+      ? signMessageEthereumKey({
```
Sign message is not running inside the same node the key is generated?

DashKash54 commented on packages/wrapped-keys-lit-actions/src/lib/common/internal/getDecryptedKey.js:
```diff
@@ -0,0 +1,20 @@
+  dataToEncryptHash,
+}) {
+  try {
+    // May be undefined, since we're using `decryptToSingleNode`
```
This should always return the decrypted data to the selected node, what is meant by this comment?

MaximusHaximus replied:
This was the existing logic, which i retained here.  My understanding is that because this code runs on all of the nodes, but the `decryptToSingleNode` call only actually returns a value on  _the one node that is chosen by backend consensus_, all of the the nodes that are executing this action that _aren't_ chosen to be 'the single node' that runs the decrypt will return undefined for this call.  As a result, we can't continue executing any further logic on that node since we don't have the privateKey in scope -- so  we exit immediately, and that node returns an empty result.  The SDK looks at the returned results from all nodes, and uses the only _non-empty result_ from the responses from all of the nodes.

DashKash54 commented on packages/wrapped-keys-lit-actions/src/lib/common/batchGenerateEncryptedKeys.js:
```diff
@@ -0,0 +1,159 @@
+    messageToSign
+      ? signMessageEthereumKey({
+          messageToSign: messageToSign,
+          privateKey: ethereumKey.privateKey,
```
`ethereumKey` will be different for each node so the `messageSignature` won't match the returned `generatedPrivateKey`

MaximusHaximus replied:
:+1: https://github.com/LIT-Protocol/js-sdk/pull/675#discussion_r1792721090

DashKash54 commented on packages/wrapped-keys-lit-actions/src/lib/common/batchGenerateEncryptedKeys.js:
```diff
@@ -0,0 +1,159 @@
+
+  const messageToSign = action.signMessageParams?.messageToSign;
+
+  const solanaKey = generateSolanaPrivateKey();
```
Same as Ethereum, different privateKey on each node

MaximusHaximus replied:
https://github.com/LIT-Protocol/js-sdk/pull/675#discussion_r1792721090 :+1:

DashKash54 commented on local-tests/tests/wrapped-keys/testBatchGeneratePrivateKeys.ts:
```diff
@@ -0,0 +1,75 @@
+      );
+    }
+
+    if (
```
Lets' add assertions to check whether the messages are signed by the `generatedPrivateKey`(s) @MaximusHaximus

MaximusHaximus replied:
I added these assertions, but per https://github.com/LIT-Protocol/js-sdk/pull/675#discussion_r1792721090 - they passed even with the existing implementation -- perhaps coincidentally.

DashKash54 commented on packages/wrapped-keys-lit-actions/src/lib/ethereum/internal/signMessage.js:
```diff
@@ -0,0 +1,39 @@
+    messageToSign,
+  });
+
+  const recoveredAddress = verifyMessageSignature({ messageToSign, signature });
```
**LATER:** Ideally this shouldn't be required and this adds an additional overhead as well, not sure by how much though. So it'll be great if we can add timings to see this

MaximusHaximus replied:
I'm happy to remove it if it's superfluous -- I only copied it in because it was in the existing implementation :+1:  Is there any case where this would fail? Usually verification of the signature is the responsibility of the receiver of the signature rather than the creator of it and it does appear that this should always work unless ethers is broken internally for some reason (at which point the consumer's verify should fail anyway?)

DashKash54 reviewed (changes requested):
@MaximusHaximus there appears to be a problem in the batch method as each node generates its own random private key and signs with it. But only one node returns the encrypted key to the user hence the encrypted key won't match the signedMessage

MaximusHaximus commented on local-tests/setup/tinny-environment.ts:
```diff
@@ -447,8 +446,9 @@ export class TinnyEnvironment {
    * @throws If there is an error sending the funds.
    */
   getFunds = async (walletAddress: string, amount = '0.001') => {
+    const privateKey = await this.getAvailablePrivateKey();
```
I needed the privateKey in scope in the finally{} block, so if I move it inside the try{}, it won't be in scope in the finally{} due to block scoping rules.  also, if this call throws an error, I don't want the finally() to run that tries to release the private key -- since we won't know what private key to release if this call throws an error

MaximusHaximus commented on local-tests/tests/wrapped-keys/util.ts:
```diff
@@ -1,7 +1,50 @@
+  EthereumLitTransaction,
+} from '@lit-protocol/wrapped-keys';
+
+const emptyLitActionRepositoryCommon: LitActionCodeRepositoryCommon = {
```
The way the fallback logic inside of the `wrapped-keys` package works, if the code for a particular LIT action is an empty string (falsey), it uses CIDs -- otherwise it uses the string as the LIT action code. It's 'empty' because it contains no code at all, for any key -- basically it resets the entire code repository so that any calls to use the LIT actions will use the defined CIDs instead of trying to use code.  For type safety it needs to be an object with the correct keys, where all appropriate values remain strings.

MaximusHaximus replied:
Keeping the values on the object always as strings keeps the type consistent so handling values in the object requires less complexity in typescript -- because values on the object are all of type `string`, we always know what type a value is without needing to do assertions against it, and the code to check if it's a falsy string to know if we should try to use any particular value as code or not is the same as if the value was 'null'  :)

MaximusHaximus commented on local-tests/tests/wrapped-keys/util.ts:
```diff
@@ -1,7 +1,50 @@
+    evm: '',
+    solana: '',
+  },
+  exportPrivateKey: {
```
You're right -- I wanted to move this into the common repository, but didn't want to ship a breaking change here which would mean shipping js-sdk v7 since we don't have independent package versioning yet :(

MaximusHaximus commented on packages/wrapped-keys-lit-actions/src/index.ts:
Am keen to do this, but am trying to ship this without any breaking changes, and the repository shape is an external API -- see https://github.com/LIT-Protocol/js-sdk/pull/675#discussion_r1792687675

DashKash54 commented on packages/wrapped-keys/src/lib/api/batch-generate-private-keys.ts:
```diff
@@ -0,0 +1,73 @@
+        return {
+          ...(signature ? { signedMessage: { signature } } : {}),
+          generatedPrivateKey: {
+            memo: generatedPrivateKey.memo,
```
Isn't this provided in the params passed by the user?

MaximusHaximus replied:
Yup -- but it's also the only identifying information that the consumer can get back from the LA that they can use to verify what they are getting back is what they sent, because they don't get the message back, and they don't know what the generated key would be -- so I added it as a convenience so that they could see in the result which entry was for which (easier debugging)

DashKash54 commented on packages/wrapped-keys/src/lib/api/batch-generate-private-keys.ts:
```diff
@@ -0,0 +1,73 @@
+          ...(signature ? { signedMessage: { signature } } : {}),
+          generatedPrivateKey: {
+            memo: generatedPrivateKey.memo,
+            id: id,
```
nit
```suggestion
            id,
```

MaximusHaximus replied:
:+1: changed

DashKash54 reviewed (changes requested):
Few more comments

MaximusHaximus commented on packages/wrapped-keys-lit-actions/src/lib/solana/generateEncryptedSolanaPrivateKey.js:
```diff
@@ -1,40 +1,32 @@
-        LIT_PREFIX + Buffer.from(solanaKeypair.secretKey).toString('hex');
-      let utf8Encode = new TextEncoder();
-      const to_encrypt = utf8Encode.encode(privateKey);
+  const { privateKey, publicKey } = generateSolanaPrivateKey();
```
:+1: good catch. This was working -- perhaps coincidentally due to internal incidental ordering of responses.
- I modified the LA to use `runOnce()` on the `processActions` call
- I added test assertions that verify the signatures -- but, see above -- they were _passing with the broken implementation anyway :(_

MaximusHaximus commented on packages/wrapped-keys-lit-actions/src/lib/ethereum/generateEncryptedEthereumPrivateKey.js:
```diff
@@ -1,37 +1,33 @@
-      const privateKey = LIT_PREFIX + wallet.privateKey.toString();
-      let utf8Encode = new TextEncoder();
-      const to_encrypt = utf8Encode.encode(privateKey);
+  const { privateKey, publicKey } = generateEthereumPrivateKey();
```
Should be GTG now - see https://github.com/LIT-Protocol/js-sdk/pull/675#discussion_r1792721090

MaximusHaximus commented on packages/wrapped-keys/src/lib/api/batch-generate-private-keys.ts:
```diff
@@ -0,0 +1,73 @@
+
+  const allowPkpAddressToDecrypt = getPkpAccessControlCondition(pkpAddress);
+
+  const { litActionCode, litActionIpfsCid } = getLitActionCodeOrCidCommon(
```
Actually, this is named accurately -- one of these two values will _always_ be falsy -- if the code exists, it will be returned and `litActionIpfsCid` will be undefined, and if the code doesn't exist then `litActionIpfsCid` will be set but the code will be falsy.

MaximusHaximus commented:
@DashKash54 Per our conversation, I've pushed commits that remove all `Lit.Actions.runOnce()` usage in the new generate function, and moved the generation of new private keys _inside_ of the runOnce blocks for existing keys.

Thinking about this change, though -- we could also normalize this behaviour and remove the runOnce usage for the _existing_ LIT actions, and configure them to use the same new 'only use response from a single node' logic.  This would improve performance for anyone using the per-network generate methods also.

FedericoAmura commented on local-tests/tests/wrapped-keys/testBatchGeneratePrivateKeys.ts:
```diff
@@ -0,0 +1,75 @@
+
+/**
+ * Test Commands:
+ * ✅ NETWORK=cayenne yarn test:local --filter=testSignMessageWithSolanaEncryptedKey
+ * ✅ NETWORK=manzano yarn test:local --filter=testSignMessageWithSolanaEncryptedKey
```
+1 because v7 does not support those networks anymore. I would prefer we document just those so it is consistent after merge and v7 release

MaximusHaximus replied:
:+1: updated this.

FedericoAmura commented on packages/wrapped-keys-lit-actions/src/lib/common/internal/getDecryptedKey.js:
```diff
@@ -0,0 +1,20 @@
+/* global Lit */
+
+export async function getDecryptedKey({
```
nit: can this be renamed to `getDecryptKeyToSingleNode`. Makes it easier to understand from a caller perspective that it can not be chosen and get `undefined`

MaximusHaximus replied:
Good idea :+1: <3

FedericoAmura commented on packages/wrapped-keys-lit-actions/src/lib/solana/internal/signMessage.js:
```diff
@@ -0,0 +1,53 @@
+import { Keypair } from '@solana/web3.js';
+import bs58 from 'bs58';
```
We can use ethers which is already in the nodes deno env instead of this

FedericoAmura replied:
What's more, we are dropping `bs58` from the SDK in v7 #662

DashKash54 replied:
As long as it's the same functionality and returns the Solana message in base58 it's fine @MaximusHaximus

FedericoAmura commented on packages/wrapped-keys-lit-actions/src/lib/solana/internal/signMessage.js:
```diff
@@ -0,0 +1,53 @@
+import { Keypair } from '@solana/web3.js';
+import bs58 from 'bs58';
```
We can use ethers which is already in the nodes deno env

Ansonhkg commented on packages/wrapped-keys/src/lib/lit-actions-client/batch-generate-keys.ts:
```diff
@@ -0,0 +1,55 @@
+    },
+    ipfsOptions: {
+      overwriteCode:
+        GLOBAL_OVERWRITE_IPFS_CODE_BY_NETWORK[litNodeClient.config.litNetwork],
```
it's an optional overwrite inside executeJs, this is a global overwrite by network set in the constants file. it was introduced when we were waiting for the datil-prod upgrade. Not sure if we still want this now, but it's an option for users to fetch from client-side instead of node-side

MaximusHaximus commented on local-tests/tests/wrapped-keys/testBatchGeneratePrivateKeys.ts:
```diff
@@ -0,0 +1,155 @@
+
+    const solanaMessageToSign = 'This is a test solana message';
+    const evmMessageToSign = 'This is a test evm message';
+    const results = await batchGeneratePrivateKeys({
```
Caught it while you were reviewing ❇️  :)

MaximusHaximus commented on packages/wrapped-keys/src/lib/lit-actions-client/constants.ts:
```diff
@@ -19,4 +19,8 @@ const LIT_ACTION_CID_REPOSITORY: LitCidRepository = Object.freeze({
 
-export { LIT_ACTION_CID_REPOSITORY };
+const LIT_ACTION_CID_REPOSITORY_COMMON: LitCidRepositoryCommon = Object.freeze({
+  batchGenerateEncryptedKeys: '',
```
Yup -- as soon as we've got :+1: on all the LIT action changes in this PR, I will run the new `sync-actions-to-ipfs` script, update this entry with the CIDs we've synced, and add a commit here that sets them

MaximusHaximus replied:
I just pushed a commit that updates these cids to be IPFS CIDs for the LA code in this PR <3 :)

MaximusHaximus commented:
@DashKash54 I merged in the new executeJs change from #677 and [updated `batchGenerateKeysWithLitAction()`](https://github.com/LIT-Protocol/js-sdk/pull/675/commits/c3102c67cd02b1d4b2fc671a205a22244cbc64c4) to set it.

This behaviour leverages the existing executeJs function behaviour to get the absolute fastest response from a node and return immediately as soon as 1 successful response is received.  We have the added benefit that if _any one request to any node_ succeeds, the call will succeed.  Basically the best of both worlds when it comes to both speed of response _and_ resilience to errors.

FedericoAmura commented on packages/types/src/lib/interfaces.ts:
```diff
@@ -522,14 +523,29 @@ export interface JsonExecutionSdkParams
+  /**
+   * idea: the number of nodes to pay for running executions
+   */
+  // numNodesToRunOn?: number;
```
Great idea. Even more important on Naga

DashKash54 commented on packages/wrapped-keys-lit-actions/src/lib/common/batchGenerateEncryptedKeys.js:
```diff
@@ -0,0 +1,147 @@
+    network,
+    generatedPrivateKey: {
+      ...generatedPrivateKey,
+      memo: generateKeyParams.memo,
```
This is good to return since we're generating multiple keys

MaximusHaximus replied:
:+1:
In theory the results coming back in the same _order_ that the array came in is 'enough', but being able to see the memo in the element you've 'zipped' together with an array you already had in your client context is pretty useful for debugging

DashKash54 reviewed (commented):
Left a few more comments

MaximusHaximus commented on local-tests/tests/wrapped-keys/testBatchGeneratePrivateKeys.ts:
```diff
@@ -0,0 +1,154 @@
+    }
+  }
+
+  const { decryptedPrivateKey } = await exportPrivateKey({
+    litNodeClient,
+    network: 'evm',
+    id: evmResult.generatedPrivateKey.id,
+    pkpSessionSigs,
+  });
+
+  const recoveredAddress = verifyMessageSignature();
+
+  const wallet = new ethers.Wallet(decryptedPrivateKey);
```
good call -- I actually took this assertion pattern from the original `testEthereumBroadcastTransactionGeneratedKey` test that you wrote, to keep things consistent with our existing tests. Should we go ahead and modify them both to avoid exporting the privateKey for either one

DashKash54 replied:
I think we could make this change as it's a minor one? @MaximusHaximus

MaximusHaximus replied:
:+1: Modified this test file -- but I was mistaken about it coming from `testEthereumBroadcastTransactionGeneratedKey` and the place it did come from was slightly different, so i left that one alone in the interest of time since this is kind of a :nit: and the existing test has been GTG

MaximusHaximus commented on local-tests/tests/wrapped-keys/testBatchGeneratePrivateKeys.ts:
```diff
@@ -0,0 +1,154 @@
+) {
+  const {
+    signedMessage: { signature },
+    generatedPrivateKey: { generatedPublicKey },
```
Sure :+1: changed this

MaximusHaximus commented on local-tests/tests/wrapped-keys/testBatchGeneratePrivateKeys.ts:
```diff
@@ -0,0 +1,154 @@
+
+    console.log('evm verify sig');
+    await verifyEvmSignature(
+      devEnv.litNodeClient,
+      results[0],
+      evmMessageToSign,
+      pkpSessionSigsSigning
```
All variables being passed into these functions are used -- if we stop using them, I'll remove them from the signature :)

verifyMessage requires { evmMessageToSign, evmResult.signMessage.signature }
export private key requires { litNodeClient, pkpSessionSigs }

DashKash54 replied:
This is what I meant by this comment: https://github.com/LIT-Protocol/js-sdk/pull/675#discussion_r1794310145 @MaximusHaximus

MaximusHaximus replied:
:+1: gotcha. Removed them after changing the method of verifying

DashKash54 commented on packages/wrapped-keys/src/lib/lit-actions-client/batch-generate-keys.ts:
```diff
@@ -0,0 +1,56 @@
+  } = args;
+
+  const result = await litNodeClient.executeJs({
+    numResponsesRequired: 1,
```
This sends to only one node or sends to all but waits on only one? @MaximusHaximus

MaximusHaximus replied:
Sends to all, but returns the _first successful response_.

This way...
1. We aren't unlucky and waiting on a node that we chose randomly that is slow
2. If _any_ nodes return a successful response, the request succeeds, instead of the request failing if the one random node we chose throws an error.

DashKash54 replied:
I think we should make this change as in our experience the nodes returns the value if they're hit, so we're reasonably confident about it and it reduce the load on the network as well @MaximusHaximus

MaximusHaximus replied:
All set!

DashKash54 commented on packages/wrapped-keys/src/lib/lit-actions-client/constants.ts:
This is the updated ones with consolidation and single node request?

MaximusHaximus replied:
Yes, but I will update them again once we have the final :+1: on the current LA code

DashKash54 replied:
Yeah those look good

DashKash54 reviewed (approved):
Looks good, approving but left few more comments & questions @MaximusHaximus

DashKash54 reviewed (approved):
Just a small suggestion for big PRs like this it'll be nice to keep the name same and raise another PR for updating the naming, makes easier to review @MaximusHaximus

MaximusHaximus commented on packages/types/src/lib/interfaces.ts:
```diff
@@ -687,7 +703,7 @@ export interface SigShare {
   bigR?: string;
   publicKey: string;
-  dataSigned?: string;
+  dataSigned?: string | 'fail';
```
Not sure about exactly why this case might happen -- but this was added in https://github.com/LIT-Protocol/js-sdk/pull/677/files

Ansonhkg replied:
`fail` is a string returned from the node if `Lit.Action.signEcdsa` failed to sign AND if the sigType is `K256`

MaximusHaximus commented on packages/wrapped-keys-lit-actions/src/lib/solana/signTransactionWithEncryptedSolanaKey.js:
No big deal -- may as well fix it while we're here and thinking about it, and it's is a trivial change, I went ahead and made it :+1 :)

DashKash54 replied:
We cna do a similar change to Etheruem as well, validate first decrypt later @MaximusHaximus

MaximusHaximus replied:
Already done <3 :0

DashKash54 commented:
> we could also normalize this behaviour and remove the runOnce usage for the existing LIT actions
We can update the exiting Lit Actions later as it's not only the Lit Action but also the client that needs to be updated with run on single node flag eg: `generatePrivateKey()`

DashKash54 commented:
>This behaviour leverages the existing executeJs function behaviour to get the absolute fastest response from any node and return immediately, as soon as any one successful response is received. 

It has a down side that all the node loads increase, we can reduce the network load by 8-10x if we run only on once
We haven't experience Lit Action throwing random errors like this so we should be confident to make this change @MaximusHaximus @Ansonhkg

MaximusHaximus commented on packages/wrapped-keys-lit-actions/src/lib/ethereum/signTransactionWithEncryptedEthereumKey.js:
No big deal -- may as well fix it while we're here and thinking about it, and it's is a trivial change, I went ahead and made it 👍 :)

DashKash54 commented on local-tests/test.ts:
Why comment it out?

MaximusHaximus replied:
Was just fixing that -- was a mistake :)

MaximusHaximus replied:
All set

MaximusHaximus commented:
> > we could also normalize this behaviour and remove the runOnce usage for the existing LIT actions
> 
> We can update the existing Lit Actions **later** as it's not only the Lit Action but also the client that needs to be updated with run on single node flag eg: `generatePrivateKey()`

OK, I'll leave them as is for now -- just FYI, the client change only involves adding a single parameter into each API function's file where they call `executeJs` -- not a big deal.

DashKash54 reviewed (commented):
Please address the remaining comments specifically the ones in which I've tagged you @MaximusHaximus

DashKash54 commented:
> OK, I'll leave them as is for now -- just FYI, the client change only involves adding a single parameter into each API function's file where they call executeJs -- not a big deal.

If you want feel free to do it now @MaximusHaximus

MaximusHaximus commented:
@DashKash54 All set

- Replaced `numResponsesRequired` with `runOnSingleNode` which sends only the request to a single, random node.
- I went ahead and updated the per-network generate LA's to use the new `runOnSingleNode` too
- I updated the CID repository with the latest CIDs

DashKash54 reviewed (approved):
LGTM!
